README.md
# VBF_Parser


history.txt
20230305
	1,can module, run with error "Session was not explicitly closed before it was destructed"
	2,nixnet module, nok to read independently without write

ztest/test_parse.py
import sys
import os
from time import sleep, time
from math import ceil
from io import BytesIO
import can

from ford.vbf import Vbf

vbf_path = 'E:\Github\MyGit\VBF_Parser\ztest/vbf\g.vbf'
print(vbf_path)
data = Vbf(vbf_path)
print("end")


ztest/test_nixnet_TR.py
# Run with PYTHON 3.9 NOT 3.10 !!!
import nixnet
from nixnet import constants
from nixnet import types
count = 1
interface2 = 'can2'
# with nixnet.FrameInStreamSession('CAN1') as input_session:
    # input_session.intf.can_term = constants.CanTerm.ON
    # input_session.intf.can_term = nixnet._enums.CanTerm.ON
    # input_session.intf.baud_rate = 125000

    # frames = input_session.frames.read(count)
    # for frame in frames:
    #     print('Received frame:')
    #     print(frame)

with nixnet.FrameOutStreamSession(interface2) as output_session:
    output_session.intf.can_term = constants.CanTerm.ON
    output_session.intf.baud_rate = 500000
    payload_list = [2, 4, 8, 2]
    id = types.CanIdentifier(0)
    payload = bytearray(payload_list)
    frame = types.CanFrame(id, constants.FrameType.CAN_DATA, payload)

    frame.payload = payload
    output_session.frames.write([frame])
    output_session.close


ford/uds.py
"""
This program is free software: you can redistribute it and/or modify it under
the terms of the GNU General Public License as published by the Free Software
Foundation, either version 3 of the License, or (at your option) any later
version.
This program is distributed in the hope that it will be useful, but WITHOUT
ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.
You should have received a copy of the GNU General Public License along with
this program. If not, see <http://www.gnu.org/licenses/>.
"""

import sys
from array import array
from time import sleep

from ford.simpleisotp import SimpleISOTP

def debug(str, end="\n"):
	print(str, end=end)
	sys.stdout.flush()

class Ecu:
	def __init__(self, can_interface="vcan0", ecuid=0, bs=0, stmin=0xff, timeout=5):
		self.ssock = SimpleISOTP(can_interface, ecuid, ecuid+0x08)
		self.ecuid  = ecuid

	def recv(self):
		return self.ssock.recv()

	def send(self, msg):
		return self.ssock.send(msg)

	def Mode9(self, data):
		self.send(bytearray([0x09] + data))
		msg = self.recv()
		return msg

	def Reset(self, val=0x01):
		self.send(bytearray([0x11, val]))
		msg = self.recv()

		if msg and msg[0] == 0x51:
			return True
		return False

	def UDSTesterPresent(self):
		self.send(bytearray([0x3e, 0x00]))
		msg = self.recv()

		if msg and msg[0] == 0x7e:
			return True
		return False

	def UDSReadDataByIdentifier(self, id):
		self.send(bytearray([0x22] + id))

		msg = self.recv()
		if msg and msg[0] == 0x62:
			return msg[3:]
		return False		

	def UDSWriteDataByIdentifier(self, id, value):
		self.send(bytearray([0x2e] + id + value))
		msg = self.recv()
		if msg and msg[0] == 0x6e:
			return True
		return False

	def UDSDiagnosticSessionControl(self, id):
		self.send(bytearray([0x10, id]))
		msg = self.recv()
		if msg and msg[0] == 0x50:
			return True
		return False

	def UDSReadMemoryByAddress(self, addr, size, aslen=0x24):
		slen = aslen >> 4
		alen = aslen & 0x0f

		self.send(bytearray([0x23, aslen]) + addr.to_bytes(alen, 'big') + size.to_bytes(slen, 'big'))
		msg = self.recv()
		if msg and msg[0] == 0x63:
			return msg[1:]
		return False

	def UDSSecurityAccess(self, num, key=[]):
		self.send(bytearray([0x27, num] + key))
		msg = self.recv()
		if msg and msg[0] == 0x67:
			if key:
				return True
			return msg[2:]
		return False

	def UDSRequestDownload(self, addr, size, fmt=0x00, aslen=0x44):
		slen = aslen >> 4
		alen = aslen & 0x0f

		self.send(bytearray([0x34, fmt, aslen]) + addr.to_bytes(alen, 'big') + size.to_bytes(slen, 'big'))
		msg = self.recv()

		if msg and msg[0] == 0x74:
			return msg[2:]
		return False

	def UDSRequestUpload(self, addr, size, fmt=0x00, aslen=0x44):
		alen = aslen >> 4
		slen = aslen & 0x0f

		self.send(bytearray([0x35, fmt, aslen]) + addr.to_bytes(alen, 'big') + size.to_bytes(slen, 'big'))
		msg = self.recv()

		if msg and msg[0] == 0x75:
			return msg[2:]
		return False

	def UDSRequestTransferExit(self):
		self.send(bytearray([0x37]))
		while True:
			msg = self.recv()
			if msg:
				if msg[0] == 0x77:
					return True

	def UDSTransferData(self, num, data=bytearray()):
		self.send(bytearray([0x36, num]) + data)
		while True:
			msg = self.recv()
			if msg:
				if msg[0] == 0x76:
					return True

	def UDSTransferDataEx(self, num, data=bytearray()):
		self.send(bytearray([0x36, num]) + data)
		while True:
			msg = self.recv()
			if msg:
				if msg[0] == 0x76:
					return msg[2:]
				else:
					return False

	def UDSRoutineControl(self, cmd, data=b''):
		self.send(bytearray([0x31]+cmd) + data)
		while True:
			msg = self.recv()
			if msg:
				if msg[0] == 0x71:
					return True

	def UDSRoutineControlEx(self, cmd, data=b''):
		self.send(bytearray([0x31]+cmd) + data)
		while True:
			msg = self.recv()
			if msg:
				return msg[2:]
	
	def getHWPartNo(self):
		return self.UDSReadDataByIdentifier([0xf1, 0x11]).decode('utf-8')

	def getPartNo(self):
		return self.UDSReadDataByIdentifier([0xf1, 0x13]).decode('utf-8')

	def getStrategy(self):
		s = self.UDSReadDataByIdentifier([0xf1, 0x88])
		if s:
			return s.decode('utf-8')
		return False

	def getCalibrationID(self):
		return self.Mode9([0x04]).hex()

	def getCVN(self):
		return self.Mode9([0x06]).hex()

	def unlock(self, level):
		seed = self.UDSSecurityAccess(level)
		if seed:
			debug("\t[+] Got seed: {}".format(' '.join(map('{:02x}'.format, seed))))
			try:
				magic = fixedbytes[self.ecuid][level]
			except KeyError as e:
				return False, "[!] Failed! No magic bytes for 0x{:x} level 0x{:02x} found. Aborting.".format(self.ecuid, level)
			debug("\t[+] Magic bytes: 0x{:x}".format(magic))
			key = keygen(seed, magic)
			keystr = ' '.join(map('{:02x}'.format, key))
			debug("\t[+] Sending key: {}".format(keystr))
			if self.UDSSecurityAccess(level+1, key):
				return True, "[+] Success!"
			else:
				return False, "[-] Failed! Aborting."
		else:
			return False, "[-] Failed to get the seed bytes for SecurityAccess 0x{:02x}".format(level)

	def SBLcall(self, addr, alen=4):
		return self.UDSRoutineControl([0x01, 0x03, 0x01], addr.to_bytes(alen, 'big'))

	def commit(self):
		r = self.UDSRoutineControl([0x01, 0x03, 0x04])
		sleep(1)
		return r

	def erase(self, addr, size, aslen=0x44):
		alen = aslen >> 4
		slen = aslen & 0x0f
		return self.UDSRoutineControl([0x01, 0xff, 0x00], addr.to_bytes(alen, 'big') + size.to_bytes(slen, 'big'))

fixedbytes = {
	0x703: {0x01: 0xfa5fc0, 0x03: 0x92c13b},
	0x726: {0x01: 0xfaa8bd, 0x11: 0x128665},
	0x727: {0x03: 0x4ad0fb},
	0x730: {0x01: 0x9b2533},
	0x731: {0x01: 0x672a70, 0x11: 0x462a71},
	0x760: {0x01: 0x582613, 0x03: 0x76807f, 0x11: 0x06316b}
}

def keygen(seed, fixed):
	challengeCode = array('Q')

	challengeCode.append(fixed & 0xff)
	challengeCode.append((fixed >> 8) & 0xff)
	challengeCode.append((fixed >> 16) & 0xff)
	challengeCode.append((fixed >> 24) & 0xff)
	challengeCode.append((fixed >> 32) & 0xff)

	challengeCode.append(seed[2])
	challengeCode.append(seed[1])
	challengeCode.append(seed[0])

	temp1 = 0xC541A9
	for i in range(64):
		abit = temp1 & 0x01
		chbit = challengeCode[7] & 0x01
		bbit = abit ^ chbit

		temp2 = (temp1 >> 1) + bbit * 0x800000 & -1
		temp1 = (temp2 ^ 0x109028 * bbit) & -1
		challengeCode[7] = challengeCode[7] >> 1 & 0xff
		for a in range(7, 0, -1):
			challengeCode[a] = challengeCode[a] + (challengeCode[a - 1] & 1) * 128 & 0xff
			challengeCode[a - 1] = challengeCode[a - 1] >> 1

	
	key = [ temp1 >> 4 & 0xff, ((temp1 >> 12 & 0x0f) << 4) + (temp1 >> 20 & 0x0f), (temp1 >> 16 & 0x0f) + ((temp1 & 0x0f) << 4) ]

	return key

ecu = Ecu(can_interface="can1", ecuid=400)
ecu.send(bytearray([0x12, 0x02]))



ford/simpleisotp.py
"""
This is by no means a full/proper ISOTP implementation. It aims to work with just the BROKEN Ford implementation.


This program is free software: you can redistribute it and/or modify it under
the terms of the GNU General Public License as published by the Free Software
Foundation, either version 3 of the License, or (at your option) any later
version.
This program is distributed in the hope that it will be useful, but WITHOUT
ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.
You should have received a copy of the GNU General Public License along with
this program. If not, see <http://www.gnu.org/licenses/>.
"""

import sys
from time import sleep
from io import BytesIO
import can

class SimpleISOTP:
	def __init__(self, can_interface, can_tx, can_rx):
		self.state = 0
		self.can_id = can_tx
		try:
			# self.bus = can.interface.Bus(bustype='socketcan', channel=can_interface, bitrate=500000, receive_own_messages=False)
			self.bus = can.interface.Bus(bustype='nixnet', channel=can_interface, bitrate=500000, receive_own_messages=False)
			
		except can.CanError as e:
			print("[!] Unable to open {}".format(can_interface))
			sys.exit(-1)

		can_filters = [{"can_id": can_rx, "can_mask": 0xfff, "extended": False}]
		self.bus.set_filters(can_filters)

	def putoncan(self, msg):
		sleep(0.002)
		return self.bus.send(msg)

	def send(self, payload):
		size = len(payload)

		if size < 8:
			self.state = 0
			data = bytearray([len(payload)]) + payload + bytearray([0x00] * (7-size))
			msg = can.Message(arbitration_id=self.can_id, data=data, is_extended_id=False)
			self.putoncan(msg)
		else:
			data = bytearray().fromhex('1{:03x}'.format(size)) + payload[:6]
			msg = can.Message(arbitration_id=self.can_id, data=data, is_extended_id=False)
			self.putoncan(msg)
			while True:
				fc = self.bus.recv()
				if fc.data[0] & 0xf0 == 0x30:
					break

			self.state = 1
			ds = BytesIO(payload[6:])

			while True:
				part = ds.read(7)
				if not part:
					self.state = 0
					break

				data = bytearray([0x20+self.state]) + part + bytearray([0x00] * (7-len(part)))
				msg = can.Message(arbitration_id=self.can_id, data=data, is_extended_id=False)
				self.putoncan(msg)
				if self.state < 0x0f:
					self.state += 1
				else:
					self.state = 0
		# self.bus.send.__closure__

	def recv(self):
		while True:
			data = self.bus.recv().data
			# print("DEBUG: {}".format(data))
			if not data:
				return None

			if self.state == 0 and (data[0] & 0xf0 == 0x10):
				self.state = 1
				size = (data[0] & 0x0f) * 0x100 + data[1]
				buf = BytesIO()
				received = 6
				buf.write(data[2:])

				data = bytearray([0x30]) + bytearray([0x00] * 7)
				msg = can.Message(arbitration_id=self.can_id, data=data, is_extended_id=False)				
				self.putoncan(msg)

			if self.state > 0 and (data[0] & 0xf0 == 0x20):
				self.state = 2

				if(received + 7 > size):
					buf.write(data[1:1+size-received])
				else:
					buf.write(data[1:])

				received += 7
				if received >= size:
					self.state = 0
					buf.seek(0)
					return buf.read()


			if self.state == 0 and (data[0] & 0xf0 != 0x30):
				size = data[0]
				return data[1:size+1]
		
# can_interface = "can1"
# ecuid = 400
# aa = SimpleISOTP(can_interface, ecuid, ecuid+0x08)
# aa.send(bytearray([0x13, 0x01]))
# aa.send(bytearray([0x11, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01]))


ford/simpleisotp2.py
"""
This is by no means a full/proper ISOTP implementation. It aims to work with just the BROKEN Ford implementation.


This program is free software: you can redistribute it and/or modify it under
the terms of the GNU General Public License as published by the Free Software
Foundation, either version 3 of the License, or (at your option) any later
version.
This program is distributed in the hope that it will be useful, but WITHOUT
ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.
You should have received a copy of the GNU General Public License along with
this program. If not, see <http://www.gnu.org/licenses/>.
"""

import sys
from time import sleep
from io import BytesIO
# import can
import nixnet
from nixnet import constants
from nixnet import types

class SimpleISOTP:
	def __init__(self, can_interface, can_tx, can_rx):
		self.state = 0
		self.can_id = can_tx
		# try:
		# 	# self.bus = can.interface.Bus(bustype='socketcan', channel=can_interface, bitrate=500000, receive_own_messages=False)
		# 	self.bus = can.interface.Bus(bustype='nixnet', channel=can_interface, bitrate=500000, receive_own_messages=False)
		# 	# self.output_session = nixnet.FrameOutStreamSession('can1')
			
		# 	# id = types.CanIdentifier(0)
		# 	# # payload = bytearray(msg)
		# 	# payload_list = [2, 4, 8, 2]
		# 	# # payload = bytearray(payload_list)
		# 	# payload = msg
		# 	# frame = types.CanFrame(id, constants.FrameType.CAN_DATA, payload)
		# 	# frame.payload = payload
		# 	# self.output_session.frames.write([frame])
		# 	# self.output_session.close

		# except can.CanError as e:
		# 	print("[!] Unable to open {}".format(can_interface))
		# 	sys.exit(-1)

		# can_filters = [{"can_id": can_rx, "can_mask": 0xfff, "extended": False}]
		# self.bus.set_filters(can_filters)

	def putoncan(self, msg):
		sleep(0.002)
		with nixnet.FrameOutStreamSession('can1') as output_session:
			output_session.intf.can_term = constants.CanTerm.ON
			output_session.intf.baud_rate = 500000
			id = types.CanIdentifier(self.can_id)
			payload = msg
			frame = types.CanFrame(id, constants.FrameType.CAN_DATA, payload)
			frame.payload = payload
			output_session.frames.write([frame])

	def send(self, payload):
		size = len(payload)

		if size < 8:
			self.state = 0
			data = bytearray([len(payload)]) + payload + bytearray([0x00] * (7-size))
			self.putoncan(data)
		else:
			data = bytearray().fromhex('1{:03x}'.format(size)) + payload[:6]
			self.putoncan(data)
			# while True:
			# 	fc = self.bus.recv()
			# 	if fc.data[0] & 0xf0 == 0x30:
			# 		break

			self.state = 1
			ds = BytesIO(payload[6:])

			while True:
				part = ds.read(7)
				if not part:
					self.state = 0
					break

				data = bytearray([0x20+self.state]) + part + bytearray([0x00] * (7-len(part)))
				self.putoncan(data)
				if self.state < 0x0f:
					self.state += 1
				else:
					self.state = 0

	def recv(self):
		while True:
			# data = self.bus.recv().data
			with nixnet.FrameInStreamSession('can1') as input_session:
				input_session.intf.can_term = constants.CanTerm.ON
				input_session.intf.can_term = nixnet._enums.CanTerm.ON
				input_session.intf.baud_rate = 125000
				count = 1
				data = input_session.frames.read(count)
				# for frame in frames:
				# 	print('Received frame:')
				# 	print(frame)
			# print("DEBUG: {}".format(data))
			if not data:
				return None

			if self.state == 0 and (data[0] & 0xf0 == 0x10):
				self.state = 1
				size = (data[0] & 0x0f) * 0x100 + data[1]
				buf = BytesIO()
				received = 6
				buf.write(data[2:])

				data = bytearray([0x30]) + bytearray([0x00] * 7)
				# msg = can.Message(arbitration_id=self.can_id, data=data, is_extended_id=False)				
				self.putoncan(data)

			if self.state > 0 and (data[0] & 0xf0 == 0x20):
				self.state = 2

				if(received + 7 > size):
					buf.write(data[1:1+size-received])
				else:
					buf.write(data[1:])

				received += 7
				if received >= size:
					self.state = 0
					buf.seek(0)
					return buf.read()


			if self.state == 0 and (data[0] & 0xf0 != 0x30):
				size = data[0]
				return data[1:size+1]
		
# can_interface = "can1"
# ecuid = 400
# aa = SimpleISOTP(can_interface, ecuid, ecuid+0x08)
# aa.send(bytearray([0x13, 0x01]))
# aa.send(bytearray([0x11, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01]))


ztest/can_frame_stream_io.py
from __future__ import absolute_import
from __future__ import division
from __future__ import print_function

import time

import six

import nixnet
from nixnet import constants
from nixnet import types


def main():
    interface1 = 'CAN1'
    interface2 = 'CAN2'

    with nixnet.FrameInStreamSession(interface1) as input_session:
        with nixnet.FrameOutStreamSession(interface2) as output_session:
            terminated_cable = six.moves.input('Are you using a terminated cable (Y or N)? ')
            if terminated_cable.lower() == "y":
                input_session.intf.can_term = constants.CanTerm.ON
                output_session.intf.can_term = constants.CanTerm.OFF
            elif terminated_cable.lower() == "n":
                input_session.intf.can_term = constants.CanTerm.OFF
                output_session.intf.can_term = constants.CanTerm.ON
            else:
                print("Unrecognised input ({}), assuming 'n'".format(terminated_cable))
                input_session.intf.can_term = constants.CanTerm.ON
                output_session.intf.can_term = constants.CanTerm.ON

            input_session.intf.baud_rate = 125000
            output_session.intf.baud_rate = 125000

            # Start the input session manually to make sure that the first
            # frame value sent before the initial read will be received.
            input_session.start()

            user_value = six.moves.input('Enter payload [int, int]: ')
            try:
                payload_list = [int(x.strip()) for x in user_value.split(",")]
            except ValueError:
                payload_list = [2, 4, 8, 16]
                print('Unrecognized input ({}). Setting data buffer to {}'.format(user_value, payload_list))

            id = types.CanIdentifier(0)
            payload = bytearray(payload_list)
            frame = types.CanFrame(id, constants.FrameType.CAN_DATA, payload)

            print('The same values should be received. Press q to quit')
            i = 0
            while True:
                for index, byte in enumerate(payload):
                    payload[index] = byte + i

                frame.payload = payload
                output_session.frames.write([frame])
                print('Sent frame with ID: {} payload: {}'.format(frame.identifier,
                                                                  list(frame.payload)))

                # Wait 1 s and then read the received values.
                # They should be the same as the ones sent.
                time.sleep(1)

                count = 1
                frames = input_session.frames.read(count)
                for frame in frames:
                    print('Received frame with ID: {} payload: {}'.format(frame.identifier,
                                                                          list(six.iterbytes(frame.payload))))

                i += 1
                if max(payload) + i > 0xFF:
                    i = 0

                inp = six.moves.input('Hit enter to continue (q to quit): ')
                if inp.lower() == 'q':
                    break

            print('Data acquisition stopped.')


if __name__ == '__main__':
    main()


ztest/simpleisotp_nixnet.py
"""
This is by no means a full/proper ISOTP implementation. It aims to work with just the BROKEN Ford implementation.


This program is free software: you can redistribute it and/or modify it under
the terms of the GNU General Public License as published by the Free Software
Foundation, either version 3 of the License, or (at your option) any later
version.
This program is distributed in the hope that it will be useful, but WITHOUT
ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.
You should have received a copy of the GNU General Public License along with
this program. If not, see <http://www.gnu.org/licenses/>.
"""
import six
import sys
from time import sleep
from io import BytesIO
# import can
import nixnet
from nixnet import constants
from nixnet import types

class SimpleISOTP:
	def __init__(self, can_interface, can_tx, can_rx):
		self.state = 0
		self.can_id = can_tx
		# try:
		# 	# self.bus = can.interface.Bus(bustype='socketcan', channel=can_interface, bitrate=500000, receive_own_messages=False)
		# 	self.bus = can.interface.Bus(bustype='nixnet', channel=can_interface, bitrate=500000, receive_own_messages=False)
		# 	# self.output_session = nixnet.FrameOutStreamSession('can1')
			
		# 	# id = types.CanIdentifier(0)
		# 	# # payload = bytearray(msg)
		# 	# payload_list = [2, 4, 8, 2]
		# 	# # payload = bytearray(payload_list)
		# 	# payload = msg
		# 	# frame = types.CanFrame(id, constants.FrameType.CAN_DATA, payload)
		# 	# frame.payload = payload
		# 	# self.output_session.frames.write([frame])
		# 	# self.output_session.close

		# except can.CanError as e:
		# 	print("[!] Unable to open {}".format(can_interface))
		# 	sys.exit(-1)

		# can_filters = [{"can_id": can_rx, "can_mask": 0xfff, "extended": False}]
		# self.bus.set_filters(can_filters)

	def putoncan(self, msg):
		sleep(0.002)
		with nixnet.FrameOutStreamSession('can1') as output_session:
			# with nixnet.FrameInStreamSession('can2') as input_session:
				output_session.intf.can_term = constants.CanTerm.ON
				output_session.intf.baud_rate = 500000

				# input_session.intf.can_term = constants.CanTerm.ON
				# input_session.intf.baud_rate = 500000

				# input_session.start()
				id = types.CanIdentifier(self.can_id)
				payload = msg
				frame = types.CanFrame(id, constants.FrameType.CAN_DATA, payload)
				frame.payload = payload
				output_session.frames.write([frame])

				# count = 1
				# frames = input_session.frames.read(count)
				# for frame in frames:
				# 	print('Received frame with ID: {} payload: {}'.format(frame.identifier,list(six.iterbytes(frame.payload))))

	def send(self, payload):
		size = len(payload)

		if size < 8:
			self.state = 0
			data = bytearray([len(payload)]) + payload + bytearray([0x00] * (7-size))
			self.putoncan(data)
		else:
			data = bytearray().fromhex('1{:03x}'.format(size)) + payload[:6]
			self.putoncan(data)
			# while True:
			# 	fc = self.bus.recv()
			# 	if fc.data[0] & 0xf0 == 0x30:
			# 		break

			self.state = 1
			ds = BytesIO(payload[6:])

			while True:
				part = ds.read(7)
				if not part:
					self.state = 0
					break

				data = bytearray([0x20+self.state]) + part + bytearray([0x00] * (7-len(part)))
				self.putoncan(data)
				if self.state < 0x0f:
					self.state += 1
				else:
					self.state = 0

	def recv1(self):
		with nixnet.FrameInStreamSession('can2') as input_session:
			input_session.intf.can_term = constants.CanTerm.OFF
			input_session.intf.baud_rate = 500000
			input_session.start()
			count = 1
			frames = input_session.frames.read(count)
			for frame in frames:
				print('Received frame with ID: {} payload: {}'.format(frame.identifier,list(six.iterbytes(frame.payload))))

	def recv(self):
		while True:
			# data = self.bus.recv().data
			with nixnet.FrameInStreamSession('can2') as input_session:
				input_session.intf.can_term = constants.CanTerm.ON
				# input_session.intf.can_term = nixnet._enums.CanTerm.ON
				input_session.intf.baud_rate = 500000
				count = 1
				data = input_session.frames.read(count)
				# for frame in frames:
				# 	print('Received frame:')
				# 	print(frame)
			# print("DEBUG: {}".format(data))
			if not data:

				return None

			# fd0 = frames[0].identifier
			# fd1 = list(six.iterbytes(frames[0].payload))
			for frame in data:
				print('Received frame with ID: {} payload: {}'.format(frame.identifier,list(six.iterbytes(frame.payload))))

			if self.state == 0 and (data[0] & 0xf0 == 0x10):
				self.state = 1
				size = (data[0] & 0x0f) * 0x100 + data[1]
				buf = BytesIO()
				received = 6
				buf.write(data[2:])

				data = bytearray([0x30]) + bytearray([0x00] * 7)
				# msg = can.Message(arbitration_id=self.can_id, data=data, is_extended_id=False)				
				self.putoncan(data)

			if self.state > 0 and (data[0] & 0xf0 == 0x20):
				self.state = 2

				if(received + 7 > size):
					buf.write(data[1:1+size-received])
				else:
					buf.write(data[1:])

				received += 7
				if received >= size:
					self.state = 0
					buf.seek(0)
					return buf.read()


			if self.state == 0 and (data[0] & 0xf0 != 0x30):
				size = data[0]
				return data[1:size+1]
		
can_interface = "can1"
ecuid = 400
aa = SimpleISOTP(can_interface, ecuid, ecuid+0x08)
aa.send(bytearray([0x13, 0x01]))
# sleep(1)
# aa.recv1()
# aa.send(bytearray([0x11, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01]))

# can_interface = "can2"
# ecuid = 400
# bb = SimpleISOTP(can_interface, ecuid, ecuid+0x08)
# rx = aa.recv()
# print(rx)